invoice.py

```diff
@@ -0,0 +1,101 @@
+import base64
+import requests
+import socket
+
+from logging import getLogger
+
+from trytond.pool import PoolMeta
+from trytond import backend
+from trytond.config import config as config_
+from trytond.modules.account_invoice_facturae import FACTURAE_SCHEMA_VERSION
+
+PIMEFACTURA_PROD = config_.getboolean('pimefactura', 'production', default=False)
+PIMEFACTURA_USER = config_.get('pimefactura', 'user', default=None)
+PIMEFACTURA_PASSWORD = config_.get('pimefactura', 'password', default=None)
+PIMEFACTURA_BASEURL = ('https://www.pimefactura.com'
+    if PIMEFACTURA_PROD else 'http://new.pimefactura.com')
+
+_logger = getLogger(__name__)
+
+def basic_auth(username, password):
+    token = base64.b64encode(f"{username}:{password}".encode('utf-8')).decode("ascii")
+    return f'Basic {token}'
+
+
+class Invoice(metaclass=PoolMeta):
+    __name__ = 'account.invoice'
+
+    @classmethod
+    def generate_facturae_pimec(cls, invoices, certificate=None,
+            certificate_password=None):
+        url = '%s/uploadinvoice' % PIMEFACTURA_BASEURL
+
+        to_write = []
+        for invoice in invoices:
+            if invoice.invoice_facturae:
+                continue
+
+            facturae_content = invoice.get_facturae()
+            invoice._validate_facturae(facturae_content)
+
+            if backend.name != 'sqlite':
+                invoice_facturae = invoice._sign_facturae(
+                    facturae_content, 'pimec', certificate, certificate_password)
+            else:
+                invoice_facturae = facturae_content
+
+            headers = {
+                'Content-Type': 'application/xml',
+                "Authorization": basic_auth(PIMEFACTURA_USER, PIMEFACTURA_PASSWORD),
+                }
+
+            data = '''
+                <UploadInvoiceRequest>
+                    <invoicetype>facturae</invoicetype>
+                    <invoicetypeversion>%(version)s</invoicetypeversion>
+                    <invoiceb64>%(invoiceb64)s</invoiceb64>
+                </UploadInvoiceRequest>
+                ''' % {
+                    'version': FACTURAE_SCHEMA_VERSION,
+                    'invoiceb64': base64.b64encode(invoice_facturae).decode('utf-8'),
+                }
+
+            try:
+                rqst = requests.put(
+                    url,
+                    data=data,
+                    headers=headers
+                    )
+            except Exception:
+                _logger.warning('Error send Pimec factura-e: %s' % invoice.rec_name)
+                continue
+
+            try:
+                if rqst.status_code == 200 or rqst.status_code == 201:
+                    to_write.extend(([invoice], {
+                        'invoice_facturae': invoice_facturae,
+                        }))
+                else:
+                    _logger.warning('Error send Pimec factura-e status code: %s %s' % (rqst.status_code, rqst.text))
```

Comment:
1.
A- In case has an error, field "invoice_facturae" from account.invoice continue to None.
B.- In case send pimec with queue, user not know

2. 
In case call first "default" method, we call twice some methods, for example: invoice.get_facturae()